url913d8566d1_AnsBalin_rbn_tgz-pJ8-MainJ8/src/Graph.java

```diff
@@ -0,0 +1,399 @@
+import java.util.*;
+
+public class Graph{
+    
+  public static final int _REACTION_ = Integer.MIN_VALUE;
+  public static final int _MOLECULE_ = Integer.MIN_VALUE + 1;
+  public static final int _NULL_ = Integer.MIN_VALUE + 2;
+  public static final int WHITE = 0;
+  public static final int BLACK = -1;
+  
+  
+  public class GraphNode{
+    
+    private Object contents;
+    private int type;
+    private int ID;
+    private int nodeIndex;
+    private String StringID;
+    private ArrayList children;
+    private ArrayList parents;
+    private int colour; // -1: black, 0: white, 1+:gray
+    
+    public GraphNode(){
+      contents = null;
+      ID=-2;
+      StringID="Null";
+      children = null;
+      parents = null;
+      type = _NULL_;
+    }
+    
+    public GraphNode(Molecule o, int nodeIndex){
+      
+      contents = new Molecule(o);
+      type = _MOLECULE_;
+      children = new ArrayList<GraphNode>();
+      parents = new ArrayList<GraphNode>();
+      colour = 0;
+      ID = o.getID();
+      StringID = o.getStringID();
+      this.nodeIndex = nodeIndex;
+      
+    }
+    public GraphNode(Reaction o, int nodeIndex){
+      
+      contents = new Reaction(o);
+      type = _REACTION_;
+      children = new ArrayList<GraphNode>();
+      parents = new ArrayList<GraphNode>();
+      colour = 0;
+      ID = o.getIntID();
+      StringID = o.getID();
+      this.nodeIndex = nodeIndex;
+      
+    }
+    
+    public int getColour(){
+      return colour;
+    }
+    
+    public int getID(){
+      return ID;
+    }
+    public int getNodeIndex(){
+      return nodeIndex;
+    }
+    public int getType(){
+      return type;
+    }
+    
+    public void print(){
+    
+      System.out.println(""+StringID);
+    }
+    
+    public ArrayList<GraphNode> getChildren(){
+      return children;
+    }
+    
+    public void colour(int i){
+    
+      colour=i;
+      
+    }
+    
+    public void addChild(GraphNode GNChild){
+      
+      children.add(GNChild);
+      
+    }
+    
+    public void addParent(GraphNode GNParent){
+    
+      parents.add(GNParent);
+    }
+    
+    //Forms directed edge from this to GN
+    public void connect(GraphNode GN){
+      
+      addChild(GN);
+      GN.addParent(this);
+      
+    }
+  } 
+  
+  private ArrayList<GraphNode> graphNodes;
+  private ArrayList<int[]> cycles;
+  private int numCycles;
+  private int visited=0;
+  private ArrayList<GraphNode> buffer;
+  
+  public Graph(){
+    
+    graphNodes = new ArrayList<GraphNode>();
+    graphNodes.add(new GraphNode());
+    numCycles = 0;
+    buffer = new ArrayList<GraphNode>();
+    cycles = new ArrayList<int[]>();
+  }
+  
+  public int checkReactionID(int IDCheck){
+    int index=-1;
+    for(int i=0; i<graphNodes.size(); i++){
+      if( (graphNodes.get(i).getID()==IDCheck) && (graphNodes.get(i).getType() == _REACTION_) ){
+        index=i;
+      
+      }
+    }
+    return index;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 122:129, 133:140.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.